core/Tracker/Db.php

```diff
@@ -261,18 +261,13 @@ public static function factory($configDb)
          */
         Piwik::postEvent('Tracker.getDatabaseConfig', array(&$configDb));
 
-        switch ($configDb['adapter']) {
-            case 'PDO\MYSQL':
-            case 'PDO_MYSQL': // old format pre Piwik 2
-                require_once PIWIK_INCLUDE_PATH . '/core/Tracker/Db/Pdo/Mysql.php';
-                return new Mysql($configDb);
-
-            case 'MYSQLI':
-                require_once PIWIK_INCLUDE_PATH . '/core/Tracker/Db/Mysqli.php';
-                return new Mysqli($configDb);
+        $className = 'Piwik\Tracker\Db\\' . str_replace(' ', '\\', ucwords(str_replace(array('_', '\\'), ' ', strtolower($configDb['adapter']))));
```

Comment:
Does this mean it's not possible to define a db adapter in a plugin (because the namespace has to be in `Piwik\Tracker\Db`? I'm wondering if getting the class name through DI or an event might be better.

Reply:
It works when you load the file where the class for this adapter is loaded yourself. DI could work as well, I just didn't want to change too much logic as in the regular DB factory it works already like this and also has the namespace `Piwik\Db...`. I reckon it's unlikely other people write their own adapters so maybe it's not too important, but could of course also change it towards DI but actually not sure if DI is already loaded at the time the DB is created? Likely it is ... but maybe not during installation etc? Not sure...

Reply:
No worries, if it's already an established method I can merge